### DISCUSSION TITLE
FTP with testcontainers

### DISCUSSION THREAD

**Discussion by rabiori** [testcontainers/testcontainers-java#4787] [Q&A] [answered]
I'm trying to start and connect to an FTP server in my tests, I have a reproducible sample and all the data here: https://stackoverflow.com/q/70274689/2924784

Any help would be greatly appreciated.

**Comment by kiview** [comment]
In the Docker command, you did open a lot of ports. You need to add all of those exposed ports to Testcontainers and get the mapped ports accordingly. This image does not seem to be container-friendly and might not work well with randomized ports, especially with regards to the ports used in passive mode in this case.

**Comment by rabiori** [comment]
I've received an answer there using the deprecated `FixedHostPortGenericContainer` - If you have any other solutions I'd be pleased to hear them

**Comment by kiview** [comment, marked as answer]
We strongly discourage usage of `FixedHostPortGenericContainer`, however, the answer on SO is still correct:
https://stackoverflow.com/a/70371790/1396362

I currently don't have a better idea for testing against an FTP container in passive mode.

**Reply by bsideup** [reply]
One can start the container first, obtain the assigned port, and then start the FTP process inside with the port already known, similar to how we did it in KafkaContainer in the past:
https://github.com/testcontainers/testcontainers-java/blob/1.16.0/modules/kafka/src/main/java/org/testcontainers/containers/KafkaContainer.java

**Reply by kiview** [reply]
I don't think this would work for passive mode in FTP, this will specifically initiate a connection on a random port in a certain range (from my understanding).
